cdap-api/src/main/java/io/cdap/cdap/api/metadata/MetadataEntity.java

```diff
@@ -94,15 +94,12 @@ public class MetadataEntity implements Iterable<MetadataEntity.KeyValue> {
     Map<String, String[][]> typesToKeys = new HashMap<>();
     typesToKeys.put(NAMESPACE, new String[][]{{NAMESPACE}});
     typesToKeys.put(DATASET, new String[][]{{NAMESPACE, DATASET}, {DATASET}});
-    typesToKeys.put(APPLICATION, new String[][]{{NAMESPACE, APPLICATION, VERSION}, {NAMESPACE, APPLICATION}});
```

Comment:
This will break backward compat for the metadata entity, as you can see the existing unit test already breaks, I think what we want to do is to hide the version from the metadata http handler response. To do that, you can introduce a codec for gson to serialize and deserialize the `MetadataEntity`, basically when it serialize the entity, you remove the version for certain type, when it deserialize, it will append -SNAPSHOT, which is the current behavior. See `MetadataCodec` on how to write the codec

Reply:
Basically after your change, the tests around `MetadataEntity` should not break

Reply:
Why do we need to serialize `MetadataEntity`? Asking because I don't find any place where we need to serialize `MetadataEntity`. In `MetadataHttpHandler` it only serializes `Metadata` objects for json responses, `metadataEntity` is send only in string responses.  
So we need to either change the toString method for `MetadataEntity` or do we actually want to remove version during `Metadata` serialization?